doc/read-the-docs-site/adr/0006-indexer-resuming-strategy.rst

```diff
@@ -45,16 +45,89 @@ Decision
   We assume that all the data contained in the indexer is consistent for all the point returned by an indexer.
 
-* We will have each indexer start the node-to-client chain-sync protocol at different points in time.
-  However, the ``Coordinator`` will make sure that all indexers are synced until a common point, and advance at the same speed (i.e. they can only request the next block once all the indexers have finished processing the current block).
+* We will start a single node chain-sync client for all indexers instead of one node chain sync client per indexer.
 
-* We will use the full set of points provided by ``resumableFromStorage`` of an indexer as resuming points for the node-to-client chain-sync protocol.
-  These points need to ordered in descending order so that protocol can priotise the selection of the latest ones.
+* We will implement the following resuming logic.
+
+  In order to start the chain-sync client, we need to provide to provide intersection points (or resuming points).
+  Intersection points represent the points from which we want to start syncing from the local node.
+  We will provide as intersection points the points returned from the following pseudo-code:
+
+  .. code-block:: haskell
+
+    -- | Query the resumable points of each indexer then select the resumable
+    -- points of the indexer that has the lowest immutable point. If two indexer
+    -- have the same lowest immutable point, select the one which has the lowest
+    -- overall point (immutable or not).
+    selectIntersectionPoints indexers = do
+      resumingPoints <- mapM queryResumingPoints indexers
+      head $ sortOn (\rps -> (maximum $ filter isImmutablePoint rps, maximum rps)) resumingPoints
+
+  Next, we need a function which processes each ``ChainSyncEvent`` returned by the chain-sync client for each indexer.
+  The pseudo-code is as follows:
+
+  .. code-block:: haskell
+
+    -- Preconditions:
+    --   * The 'initialLastSyncPoints' param should be sorted in ascending order on SlotNo
+    --   * Each element of the 'initialLastSyncPoints' param should satisfy the property 'getSlotNo point == getSlotNo (point + 1) + 1'.
+    applyChainSyncEvent initialLastSyncPoints previousChainSyncEventPoint event indexer = do
+      case event of
+        Resume point -> do
+          maybePoint <- find (\p -> getSlotNo p == getSlotNo point) initialLastSyncPoints
+          case maybePoint of
+            Nothing ->
+              pure () -- Do nothing
+            Just _ ->
+              update initialLastSyncPoints $ filter (\p -> getSlotNo p > getSlotNo point)
+        RollBackward point -> do
+          rewind point indexer
+        RollForward block point -> do
+          case initialLastSyncPoints of
+            -- Processed all initial last sync points. The indexer is in sync.
+            -- So we just index the block.
+            [] -> do
+              insert block indexer
```

Comment:
Implementation details: we can probably check if the tail of `initialLastSyncPoints` is empty at the end of a `RollForward` event rather than waiting on the next call.

Reply:
I don't see how you want to write this though

Reply:
Oh right, there's no recursive call in `applyChainSyncEvent` nevermind then, the logic is correct, that's the most important part.